### PR TITLE
Global plugin setting

### PR DESCRIPTION
**Pull request by ppwadhwa**
In the plugin installation dialog, you can select to enable/disable a plugin globally.  This PR will save this setting and propagate that setting to all the hooks.  

# Description

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by tlambert03** [comment]
thanks for getting this going @ppwadhwa.  My first thought here is that we might not want to store this information in the `HookImplementation.enabled` attribute, but rather on the plugin manager itself, and the model here would be a simple list of strings with disabled plugins.  This might require a little modification to the HookCaller object in `napari-plugin-engine`, (which would then just omit plugins that were disabled when iterating with `get_hookimpls()`), but if that's what it takes to keep the model clean, I think that's what we should do.

Note: that approach would then make it really easy to simply enable/disable a plugin without losing the enabled/disabled state of individual hook implementations provided by that plugin

**Comment by ppwadhwa** [comment]
@tlambert03 thanks so much for looking at this!  I have read your suggestion a few times and just want to make sure I understand---  so I will try to re-word and hopefully it means the same thing. :)
In `_defaults.py` we will save `plugin_state` to be a list of strings of disabled plugins.  The plugin manager will know of this list.
When you call `hook_caller.get_hookimpls()`, you will only return the plugins that are enabled (this will need to be modified in `napari-plugin-engine`).  

Is that right?

I'm still trying to work through how this will affect the plugin_sorter dialog.  I want to make sure that displays properly.  I am not sure it will know of the state of the individual plugins in each hook if its not set in `HookImplementation.enabled` at least as the way things are right now.    Right?  or am I missing something?  I feel like I've gone round and round trying to keep this all straight that I'm sure I'm overthinking or missing something.

Thanks for your patience, @tlambert03 !

**Comment by ppwadhwa** [comment]
I suppose we can add something in the plugin_sorter to enable or disable a checkbox based on the list in settings.plugins.plugin_state.  I do see what you're saying about not losing the state in the individual hooks.

**Comment by tlambert03** [comment]
> Is that right?

yep, I'd say you understood perfectly :)  note, I'm not 100% sure that my approach is good!  but you definitely understood me :)

> how this will affect the plugin_sorter dialog.

yeah, that's a good question and will probably determine whether the above suggestion will work.  I guess we should decide how we want that to look.  If we still want a disabled plugin to show all of it's hook implementations in the plugin sorter, then rather than changing the way `get_hookimpls()` is called, we may need to either

- modify the `enabled` attribute to be more of a computed property:
```python
@property
def enabled(self):
    # something like:
    return self._enabled and not self.plugin in plugin_manager._disabled_plugins
```

or, we add a field in _addition_ to `enabled` that somehow captures that the whole plugin is disabled.

some UI questions to consider in the sorter:

- what should happens when you click on a hook implementation from a disabled plugin?  Does it enable that plugin, or is it totally unclickable?  
- If it enables it, what about the other hook impls from the same plugin that were enabled before the plugin itself was disabled?

if we just say that a disabled plugin doesn't show _anyhere_, it all gets easier (it's just a plugin that is installed in the environment, but for all intents and purposes doesn't exist other than to show up in the "installed" plugins list)

thoughts?

**Comment by ppwadhwa** [comment]
I feel less inclined to remove the plugins from everywhere when it is disabled globally.  It may be nice to see that there are options available you could have forgotten about.  I know it could make things easier, but I do like having them visible.  

If we disable the checkboxes so they are unclickable, that will be easier in terms of not having to deal with what to do about global state when someone clicks a checkbox in the plugin_sorter, but it may be kind of annoying to the user if you're in the preferences dialog and want to enable a plugin from there, but have to close and go to the other plugin dialog.

this leaves allowing the user to click on a box, but then having to change the global state.  Does it just go back to enabling the plugin...so this will reset all the hooks to what they were?

**Comment by goanpeca** [comment]
Hi @ppwadhwa thanks for working on this and @tlambert03 for the suggestions.

I agree with moving the enabled state to `napari-plugin-engine` it makes sense, but since it touches that other repo I would leave that or a separate PR to add there? Maybe open a new issue to track that pending work?

In the meantime here is my reasoning on the global state of plugins:
1. If you disable a plugin globally, disable the respective hooks on the plugin sorters, that is grey them out and do not allow to click the checkbox. This makes it clearer to an user that a plugin is disabled globally. I think removing the entries from the list would make it less clear to users what is going on.
1. I suggest we change the current way the plugin sorter darkens when unselected to looking the same as selected, just with the change on the checkbox, and use the darkening to indicate the disabled state (when the plugin is disabled globally)
1. If you enable a plugin make sure the previous enabled/disable state of the hooks is there.
1. This raises an issue that sorting plugins in the preferences does not provide a way to enable/disable them globally so I would probably also add (in a separate PR) the list of plugins in the preferences dialog with the checkbox to enable and disable them globally.

Thoughts @napari/core-devs ?

---

I have given a bit of thought on how the plugin dialog plays now with the preferences and I will open a new issue to go over these ideas.

**Comment by sofroniewn** [comment]
Hi all, catching up on this discussion now.

> I agree with moving the enabled state to napari-plugin-engine it makes sense, but since it touches that other repo I would leave that or a separate PR to add there? Maybe open a new issue to track that pending work?

I like this idea too. I'm not sure @tlambert03 how best to do that. Maybe you can make a quick PR over there, not sure how much work it would be.

> If you disable a plugin globally, disable the respective hooks on the plugin sorters, that is grey them out and do not allow to click the checkbox. This makes it clearer to an user that a plugin is disabled globally. I think removing the entries from the list would make it less clear to users what is going on.

That's definitely very reasonable, but I also wouldn't object though if the hooks on the sorters disappeared entirely. One of the motivations for wanting a global enable/ disable was to easily check if a plugin is causing bad things to happen on import and be able to turn it off without uninstalling it. For example I think it's possible to have a plugin import the viewer and do weird monkey patching things that would cause unexpected errors in napari. Rather than having to uninstall every plugin to figure out which one is causing problems being able to just disable them seems much nicer. I think as little plugin code as possible should be run for a disabled plugin - if possible i wouldn't ever import anything from it.

Another motivation is that sometimes we've had issues with plugin imports being very slow and one could have many slow plugins that you'd want to disable until you need them, so again - I'd trade the ability not to import anything from a plugin with the ability to see grayed out things in the sorter, but I'm not sure of the details here. Maybe @tlambert03 can expand on this, if my ideas are way off or if this will be possible with the grayed out approach too.

> If you enable a plugin make sure the previous enabled/disable state of the hooks is there.

Yes I definitely agree with this, I see the global enabled/disable as a different thing, more of a master switch.

> This raises an issue that sorting plugins in the preferences does not provide a way to enable/disable them globally so I would probably also add (in a separate PR) the list of plugins in the preferences dialog with the checkbox to enable and disable them globally.

I guess, we could synchronize state with the checkbox here 

Screen Shot 2021-04-22 at 9 19 27 AM

Or we could only have enabled/ disabled in preferences? I don't mind starting with it only in the plugin installer or having it duplicated.

Hope this additional perspective helped!!

**Comment by goanpeca** [comment]
> Or we could only have enabled/ disabled in preferences? I don't mind starting with it only in the plugin installer or having it duplicated.

I actually would favor this :)

**Comment by tlambert03** [comment]
> I like this idea too. I'm not sure @tlambert03 how best to do that. Maybe you can make a quick PR over there, not sure how much work it would be.

can do

> Rather than having to uninstall every plugin to figure out which one is causing problems being able to just disable them seems much nicer. I think as little plugin code as possible should be run for a disabled plugin - if possible i wouldn't ever import anything from it.

Yeah, I agree with this.  It's conceivable that an "otherwise useful" package might write a napari plugin that for some reason really just has issues.  If those issues have anything to do with import, then if we try to show all the hooks available in a disabled plugin, it might essentially mean that you _can't_ even have that other package in the same environment...  So I think i'm still slightly tilting towards favoring just making it all invisible if it's disabled (while somehow making it clear what has been disabled)

> I'd trade the ability not to import anything from a plugin with the ability to see grayed out things in the sorter

me too. That's possible with the `plugin_manager.is_blocked()` method which gets used during discovery [here](https://github.com/napari/napari-plugin-engine/blob/59a172944ffb7e435af75b66dbc0019a1f2c48f6/napari_plugin_engine/manager.py#L239).  So as long as we had a list of "disabled" plugins (as strings) we can block them during registration

> Or we could only have enabled/ disabled in preferences? I don't mind starting with it only in the plugin installer or having it duplicated.

yeah, I'm also open to whatever here.  Duplicating or not duplicating... both ok with me (duplicating is probably easier at first?)

**Comment by tlambert03** [comment]
Hi @ppwadhwa,

following up on the conversation from the meeting yesterday, here's what I think would be a good first step.  This doesn't preclude us doing more interesting things with the visual presentation of hookimpls from disabled plugins in the future, but it at least gets us started with a basic model (and settings for) of deactivated plugins:

1. add a set-of-strings of disabled plugins to the Settings model
```python
class PluginsSettings(BaseNapariSettings):
    disabled: Set[str] = Field(default_factory=set)
```

2. in `plugin.__init__`, immediately after `plugin_manager = PluginManager()` update the plugin manager with `plugin_manager._blocked.update(SETTINGS.application.plugins.disabled)`.  (after #2535, we'll just put that in our new subclass, but that's fine for now).   That will prevent things from getting imported _at startup_.

3.   To disable an already-registered plugin (i.e. when someone unchecks the box), you can use 

```python
plugin_manager.set_blocked('plugin_name', True)  # True is the default
```

and that will also call `plugin_manager.unregister()` which remove the plugin from all of the hookcallers (so we don't actually need to worry about modifying `get_hookimpls()` upstream at all.

We need to think a bit about all the places that plugins have been "added" and probably clean it up a bit (but that can potentially come in a later PR).  One thing that comes to mind is that you probably want to `del napari.plugins.dock_widgets['plugin_name']` and `plugins.function_widgets['plugin_name']` and `_sample_data['plugin_name']` ...

4. To re-enable a plugin, when the box is re-checked... it should just be:

```python
plugin_manager.set_blocked('plugin_name', False)
plugin_manager.discover()
```

For the sake of this PR, I'd be fine not yet adding a new widget to the preferences window (and just hooking up the checkboxes in the installer window). but if you want to tackle that too that's fine!

5. We obviously then need to write out the set at `plugin_manager._blocked` to `SETTINGS`.  Not immediately sure how you want to hook up that event... but you can perhaps just re-implement `set_blocked()` on our plugin manager subclass to update the SETTINGS set?
 
hope that was clear and unblocks you for now.... but let me know if not!

**Comment by ppwadhwa** [comment]
thanks, @tlambert03 !  It all makes sense right now.  I will let you know if I have any trouble implementing this.  

I will leave the new widget in preferences for a follow up PR.  I think that will make things more simple for me---focusing on one thing at a time.  :)

**Comment by ppwadhwa** [comment]
I'm getting close here.  I am trying to see a change when I remove the disabled plugins from the various lists--- `napari.plugins.dock_widgets['plugin_name']` for example.  I want to see the menu update and I'm finding it hard to get it do that.  I've tried refresh, repaint, I tried to delete the menu and recreate it.  I will look again in the morning, but if you have any pointers @tlambert03 I'd appreciate it!

**Comment by tlambert03** [comment]
> I'm getting close here. 

awesome!  here's a preview for others:

https://user-images.githubusercontent.com/1609449/116712024-6b4d7800-a9a1-11eb-8f2f-dc055db47181.mov

> I am trying to see a change when I remove the disabled plugins from the various lists--- `napari.plugins.dock_widgets['plugin_name']` for example. I want to see the menu update and I'm finding it hard to get it do that. 

yeah, I can see that this isn't really set up yet to be that dynamic.  That menu is created [here](https://github.com/napari/napari/blob/f49e81a15c024a0a67124907f5be70e9185e5eaa/napari/_qt/qt_main_window.py#L719-L745), and isn't really set up to be dynamic.  

Honestly, I think you just shouldn't worry about dock widgets or samples for this PR.  The only error we risk here is someone disabling a plugin, and then _in the same viewer_ trying to use that plugin, at which point they'll get an error like `KeyError: "Plugin 'skimage.filters' does not provide any dock widgets"` ... so I think they should be able to figure that one out for themselves.

Ultimately, I think we might want to make our `PluginManager` instance a little bit more "evented" and have all of these menus and things listen for changes (rather than trigger them manually when needed).  All of this reorg should wait until after #2535.

File: napari/utils/settings/_defaults.py
Comment (tlambert03) [inline napari/utils/settings/_defaults.py]: unless you ran into issues that necessitated this type alias, i think this one reads easier just as `Set[str]` (without having to go find the alias to see what a `DisabledSet` type really is)

File: napari/_qt/widgets/qt_plugin_sorter.py
Comment (tlambert03) [inline napari/_qt/widgets/qt_plugin_sorter.py]: did this one not work as `widg.on_changed.connect(self.on_changed.emit)`?

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [inline napari/_qt/qt_main_window.py]: I'm confused why this is required if you also have `plugin_manager._blocked.update(...)` in `plugins.__init__`?

File: napari/_qt/dialogs/preferences_dialog.py
Comment (tlambert03) [inline napari/_qt/dialogs/preferences_dialog.py]: I'm a little concerned here that `PreferencesDialog` is starting to concern itself with some specific details of the plugin manager.  It feels a little surprising to me that the value emitted by `valueChanged` would _not_ actually be related to the value that recently changed... but rather would go do a deepcopy of the `SETTINGS.plugins.disabled_plugins` and emit that.  (I can see you're doing it because you wanted that value for the slot that handled the signal... but just flagging that at first read it feels unrelated).

Also a little concerned that the "source of truth" here is getting spread away from the plugin manager.  i.e. if one wants to know what plugins are disabled, should they go to SETTINGS?  or should they go to plugin_manager?  (plugin manager would have been my assumption, which would tell `SETTINGS` when it has changed).

again, a lot of this would be easier for you if plugin manager itself was evented.  But just wanted to note that it kinda feels like this is going in a direction of being a bit harder to follow...

File: napari/utils/settings/_defaults.py
Comment (ppwadhwa) [reply napari/utils/settings/_defaults.py]: The first thing I tried didn't work, but it was probably just a small thing that I didn't try too hard to fix--- I quickly pivoted to what I knew worked and didn't look back.  I changed it to this simple way and it seems fine.

File: napari/_qt/widgets/qt_plugin_sorter.py
Comment (ppwadhwa) [reply napari/_qt/widgets/qt_plugin_sorter.py]: You're right, this was not needed.  Thanks for catching that!  I think it was left over from something I was trying.  I went through and cleaned up all the signals.

File: napari/_qt/qt_main_window.py
Comment (ppwadhwa) [reply napari/_qt/qt_main_window.py]: that's right, you don't need both.  I should have cleaned this up.  I guess thinking about it though, I would prefer to keep this one and not the one in the `plugins.__init__ `, because I had to import SETTINGS in the init file just for that one line.  I would prefer not to do that, however, if you believe it belongs in the init file, I will keep it that way.  I'm not sure how much that matters, but I was thinking it was better not to import SETTINGS more places than needed.  What do you think @tlambert03 ?

File: napari/_qt/dialogs/preferences_dialog.py
Comment (ppwadhwa) [reply napari/_qt/dialogs/preferences_dialog.py]: @tlambert03  I think for me it was easier for me to know which plugins to re-enable when I clicked on `restore` defaults by taking the set stored in SETTINGS (that deep copy I was emitting) and unblocking those.  I agree it would be cleaner to go to the plugin manager to find out what is disabled and re-enable.  I can make that change so this is not as confusing.  Is there a quick way to re-enable everything?  I will look at it more to figure it out, but if you can give direction, I'd appreciate it!  

I think the source of truth *should* be the plugin manager.  That actually enables/disables plugins. At this point, SETTINGS really just needs to know the current state of things to be able to persist across sessions.  Is that what you're thinking?

Comment (ppwadhwa) [reply napari/_qt/dialogs/preferences_dialog.py]: you know what?  I think I can just go through each plugin and set blocked as false.  there aren't that many.  that's my plan unless there is a better way.

Comment (ppwadhwa) [reply napari/_qt/dialogs/preferences_dialog.py]: I've changed it to not rely on the settings value to reset the plugins.

**Comment by ppwadhwa** [comment]
@tlambert03 thank you so much for taking the time to look this over and give some really helpful feedback!  I have made a few adjustments noted above.  I know one thing I haven't addressed yet is setting the initial disabled plugins on start up.  I had a question above about that (whether to keep it in the `qt_main_window` file or the `plugins.__init__` file).  

I'll be on the lookout for your thoughts.  Thank you so much!

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [reply napari/_qt/qt_main_window.py]: I think I'm personally a bit less concerned with imports (short of circular import problems of course) than I am about spreading the business logic for plugins across multiple modules.
Since it's really important that we update the `plugin_manager._blocked` list _before_ discovery happens (lest all of those plugins will be imported), I definitely prefer this update to happen over in `plugins.__init__`.   

Also, though we definitely don't use it often, it's important to kinda keep the "headless" use case of napari in the back of our minds.  That is: there should be no assumption that a `QMainWindow` is even going to be created when using napari (in fact, the majority of our test suite [can pass even if Qt is not installed](https://github.com/napari/napari/runs/2482817391) in the environment).  So we also shouldn't rely on the creation of a Qt window for something as fundamental as blocking package discovery.  (same goes for call_order really... imagine a script that wants to use napari reader plugins, but doesn't really care whether a main window is created or not)

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [inline napari/_qt/qt_main_window.py]: tell me about what's going on here?  `iter_available()` is basically going to look through every package in the environment for the napari `entry_point` (and then it will happen once again in `plugin_manager.discover()` below... 

Are you just trying to unblock all plugins?  If so, you can just call `plugin_manager._blocked.clear()`

File: napari/_qt/dialogs/qt_plugin_dialog.py
Comment (tlambert03) [inline napari/_qt/dialogs/qt_plugin_dialog.py]: why not just `widg.on_changed.connect(self.state_changed.emit)`

File: napari/_qt/dialogs/qt_plugin_dialog.py
Comment (tlambert03) [inline napari/_qt/dialogs/qt_plugin_dialog.py]: why not just `self.installed_list.on_changed.connect(self.state_changed.emit)`?

File: napari/_qt/dialogs/qt_plugin_dialog.py
Comment (tlambert03) [inline napari/_qt/dialogs/qt_plugin_dialog.py]: This feels like a potential dilution of the "single source of truth".  Part of the reason that this dialog didn't need to have any signals before, is that nobody else is was "subscribing" to it (it's just a "view" onto the state of the plugin manager ... which is the _real_ source of truth here).

Currently, in this PR ... we have `QtPluginDialog` emitting a `state_changed` signal that the `QMainWindow` is listening for... and then the `QMainWindow._update_plugins` handles updating the `plugin_manager`.  Why involve QMainwindow at all here?  Why not just have the PluginDialog itself do it (it's primary purpose is to interact with the plugin manager anyway).  Similarly the `on_changed` -> `_update_call_order` signal seems to be involving the QMainWindow in something it doesn't need to be involved in.

In general, the less logic we can put on the main window, and the more we can put into temporary dialogs that pop up, do their business, and go away without needing to create connections across multiple objects, the better.

Comment (goanpeca) [reply napari/_qt/dialogs/qt_plugin_dialog.py]: > In general, the less logic we can put on the main window, and the more we can put into temporary dialogs that pop up, do their business, and go away without needing to create connections across multiple objects, the better.

I would say the less logic we put on temporary dialogs the better, as you said, they are just viewers and should be as "dumb" as possible.

Comment (tlambert03) [reply napari/_qt/dialogs/qt_plugin_dialog.py]: Something has to modify the model (the plugin manager)... and I don't see why that should be the main window?

Comment (goanpeca) [reply napari/_qt/dialogs/qt_plugin_dialog.py]: Cause the main window is the closest to a "controller" in the MVC paradigm.

Comment (tlambert03) [reply napari/_qt/dialogs/qt_plugin_dialog.py]: strongly disagree on this one.  First, we don't _have_ an MVC (we have a MV, like most [Qt patterns](https://doc.qt.io/qt-5/model-view-programming.html)).  Second, while the general concept of a "main window" might be a reasonable object to act as a controller... that's just not how napari is setup.  our main window is _very_ dumb.  Currently, you can open up a QtPluginDialog without a main window at all, interact with the general state of the plugins, and you'll be fine.  No state becomes inconsistent.  Easy to test in isolation. etc...  
With this change, you _must_ have a QMainWindow to test or use any of it.  There's really nothing else on the main window that makes it a good candidate for this.  Looking at the `QMainWindow._update_plugins` method and you'll see it basically doesn't use the `self` object at all.  It's just a plain function that _happens_ to live on the QMainWindow...

Comment (goanpeca) [reply napari/_qt/dialogs/qt_plugin_dialog.py]: Ok 👍🏼

Comment (ppwadhwa) [reply napari/_qt/dialogs/qt_plugin_dialog.py]: Hi @tlambert03, thanks again for your feedback and explanations.   I moved all the changes to the plugin manager inside the plugin dialog and off the main window.  This includes both changes to state and call order.  Please have a look and let me know what you think.

**Comment by ppwadhwa** [comment]
@tlambert03 Question about resetting preferences --  Currently, when clicking `restore defaults` in the preferences menu, all SETTINGS are restored.  I have a signal passed to the main window that will reset the plugin manager in the main window.  Is this ok?  I think above you mentioned not wanting the preferences dialog to concern itself directly with the plugin manager (or at least that's how I interpreted it), but then I also know now that we want to keep the main window dumb.  What are your thoughts on how to approach this?   thanks!

**Review by sofroniewn — COMMENTED** [review]
This is looking good. One thing that I caught was that right now disabling a plugin doesn't remove it from the add_dock_widget drop down, can that be done? Selecting it then leads to an error, but would be better if it wasn't there

Screen Shot 2021-05-03 at 7 33 51 PM

**Comment by ppwadhwa** [comment]
>One thing that I caught was that right now disabling a plugin doesn't remove it from the add_dock_widget drop down,

@sofroniewn yes, that's right.  i tried to look into getting that menu to update, but wasn't getting anywhere.  

Here is @tlambert03 's response from above regarding the menu updates:

> yeah, I can see that this isn't really set up yet to be that dynamic. That menu is created here, and isn't really set up to be dynamic.
>Honestly, I think you just shouldn't worry about dock widgets or samples for this PR. The only error we risk here is someone disabling a plugin, and then in the same viewer trying to use that plugin, at which point they'll get an error like KeyError: "Plugin 'skimage.filters' does not provide any dock widgets" ... so I think they should be able to figure that one out for themselves.

>Ultimately, I think we might want to make our PluginManager instance a little bit more "evented" and have all of these menus and things listen for changes (rather than trigger them manually when needed). All of this reorg should wait until after #2535.

I get the impression that we can work on this later, with more events in the PluginManager.  What are your thoughts?  I haven't tried again since Talley's comment, but I can look into this more if there is a way for it to be done.

**Review by sofroniewn — APPROVED** [review]
Ah ok, apologies for missing that. I'm approving now then and think we can merge as is, but if we do then @ppwadhwa can you create an issue with that info so that we know not to loose track of it and have it there in case anyone else finds it too. We can keep improving this

**Comment by ppwadhwa** [comment]
@sofroniewn Sure, I can create that issue. :)

how close is #2535 to being merged?  I think I will need to change some things in this PR to fit with those changes.

**Comment by tlambert03** [comment]
very...  just cleaning a couple things

**Comment by ppwadhwa** [comment]
@tlambert03 @sofroniewn @goanpeca I have included the changes from #2663 in this PR.   Please take a look!  I may have put some things in the wrong place.  
I have also included changes to update the dock widgets and sample data menus when a plugin is enabled/disabled.   


https://user-images.githubusercontent.com/54282105/117847513-05d27480-b248-11eb-83b6-aeb539e6668b.mov

**Review by sofroniewn — APPROVED** [review]
> I have also included changes to update the dock widgets and sample data menus when a plugin is enabled/disabled.

Yay! I just took this for a spin, all behavior works great for me. Code looks great too. Maybe @tlambert03 wants to take a final look and then we can give it a merge.

Great work here!!

**Comment by tlambert03** [comment]
looking good pam.  just had a quick question/thought as I'm reading this...

One of my hopes for adding more events to the plugin manager was to allow _it_ to be the source of truth that things like menus can listen to for updates.  I was surprised to see that #2663 wasn't used here... 
What I was imagining is that rather than passing the signal that "something has changed in the plugin manager" via a whole chain of Qt objects:

1. user clicks on a checkbox in the plugin menu
2. `_emit_state` callback emits `PluginListItem.on_changed.emit()`
3. signal caught and re-emitted by `QPluginList.on_changed.emit()`
4. signal caught, processed and re-emitted by `QPluginDialog`
5. signal caught and processed by `QMainWindow._update_menus`
6. that then clears the dock widget menu and calls `_fill_dock_widget_menu`
7. `_fill_dock_widget_menu` then goes back to the `plugin_manager` to query the state...

why not just directly update the plugin manager and make sure that the necessary views are listening.  In this case, the "view" is the dock widget menu:

1. user clicks on checkbox in plugin menu
2. `PluginListItem` _itself_ updates the `plugin_manager.set_blocked()`
3. QMainWindow (and anything else) gets notified by listening to `plugin_manager.events.disabled`
4. It can then add or remove the _one_ item it needs to based on the event...

**Comment by ppwadhwa** [comment]
@tlambert03 thanks Talley.  I get stuck in a mindset and its hard for me to back out.  I will work on getting this into place.

**Comment by tlambert03** [comment]
if you're starting to fee like "oh FFS..." on this PR, I'm also happy to pair up and do part of it!  don't necessarily need to noodle on this one too long :)

**Comment by goanpeca** [comment]
>if you're starting to fee like "oh FFS..." on this PR, I'm also happy to pair up and do part of it! don't necessarily need to noodle on this one too long :)

Language xD

**Comment by ppwadhwa** [comment]
@tlambert03 Did you push your code from our discussion yesterday somewhere?  Or can you push it here, please?  Thanks for all your help!!

**Comment by tlambert03** [comment]
sorry!  got caught up with other stuff after our meeting... lemme clean up and push

**Comment by ppwadhwa** [comment]
@tlambert03 no problem!  Its not a big rush, I just didn't want it to get forgotten.  :). thank you!

**Comment by tlambert03** [comment]
see https://github.com/ppwadhwa/napari/pull/1

**Comment by ppwadhwa** [comment]
thanks, @tlambert03 !

**Comment by ppwadhwa** [comment]
@tlambert03 @sofroniewn I have merged Talley's changed into my branch/PR.  I think the summary of the state of things now is that everything is going through the plugin manager.  So, if a plugin state is changed, the plugin manager will be updated directly and if the plugin manager has a change, then all relevant widgets/dialogs that need information from the PM will be listening to the PM for changes. 

The only thing that is not working now is preserving the call order for the plugin sorter if a plugin is disabled and then re-enabled.  It will just insert the plugin fully re-enabled in all hooks back into the plugin sorter.  If you close down napari and don't touch the plugin sorter, and then re-open your old call order will be in tact.  Adjusting the call order in the sorter will save the new call order in SETTINGS.  

@tlambert03 You said you were going to accommodate more on the plugin manager to preserve call order?  I know you had some thoughts to share on this breaking change for preserving call order...

Also, @tlambert03 You said something about not knowing why there were new signals on the preferences dialog---  I added those in order to reset the plugin states for all plugins back to True when someone selects the "Restore defaults" button in the preferences dialog.  I believe I moved that action out to the qt main window because the preferences dialog knew nothing of the plugin manager, and I wanted to update the PM outside of it.  I think in the future, if we move the plugin state options to the preferences dialog, this won't be necessary.  

Let me know how y'all want to proceed.  thank you!!

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [inline napari/_qt/qt_main_window.py]: quick comment here on what "work" means?  (like what would happen if this wasn't here?)

**Review by tlambert03 — APPROVED** [review]
>  if a plugin state is changed, the plugin manager will be updated directly and if the plugin manager has a change, then all relevant widgets/dialogs that need information from the PM will be listening to the PM for changes.

🎉

> The only thing that is not working now is preserving the call order for the plugin sorter if a plugin is disabled and then re-enabled. It will just insert the plugin fully re-enabled in all hooks back into the plugin sorter. 

that's totally fine with me and can be addressed in a follow-up.

> @tlambert03 You said you were going to accommodate more on the plugin manager to preserve call order? I know you had some thoughts to share on this breaking change for preserving call order...

yeah, need to think about exactly where that info should go.  but let's put this PR to bed first and then go there :)

> I think in the future, if we move the plugin state options to the preferences dialog, this won't be necessary.

Yeah, I agree... this looks like another case of the main window listening to signals coming from the Preferences dialog in order to update the plugin manager, when the preferences could just be updating plugin manager directly...  But let's leave that for another PR

This is good by me!

File: napari/_qt/qt_main_window.py
Comment (ppwadhwa) [reply napari/_qt/qt_main_window.py]: @tlambert03 thanks for catching that!  It is no longer needed.  I think I used to pass in SETTINGS.plugins.call_order somewhere.  I removed this part.

**Comment by ppwadhwa** [comment]
@tlambert03 I requested your review again because I wanted to make sure you saw that I removed that code snippet.  Thanks for asking about it, because its no longer necessary.  :)

**Review by sofroniewn — APPROVED** [review]
Functionality works great, code looks good to me! I'm not sure why tests are failing. One error is this

```
  E       AttributeError: 'PluginWidget' object has no attribute 'hook_list'
```

Do we need to bump a napari_plugin_engine dependency or something else?

Anyway I'm approving now, but this shouldn't be merged till all tests are passing. Great work @ppwadhwa and thanks for helping too @tlambert03 - I know this was a long one, but I'm already excited to use it to turn off some of the crazy things I have in my environment :-)

**Comment by ppwadhwa** [comment]
I was getting this error on my computer until I reset Napari to default settings, then it passed.  I'm not sure why its failing here, but I thought I'd mention this info in case it leads to any insight.  I cannot figure out what is going on.

**Comment by tlambert03** [comment]
got it... the issue was that [this line](https://github.com/napari/napari/pull/2565/commits/ca8b5570f31b6401421c8a596fd1a4a3d9cc9153#diff-11272e8ec266cf7d9e6c3c43337de19068b409531914e65ec6f9213773e13a1bR305) was triggering lazy plugin discovery, which triggered some of the events we had just connected, before the widget had completed initializing.  

The reason it manifested in the json_schema form  `PluginWidget` is because that subclasses `QtPluginSorter`. 

tests are passing locally now, if they pass here (not counting threading segfaults) lets merge

**Comment by tlambert03** [comment]
two unrelated segfaults now.  merging.  thanks @ppwadhwa!

**Comment by ppwadhwa** [comment]
@tlambert03 thank you for looking at this!!  We did it!

**Pull request merged**
